### PR TITLE
tiff need check TIFFTAG_SAMPLEFORMAT, should not always use unsigned.

### PR DESCRIPTION
**Pull request by Harvey-Huang**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=linux,win64,mac,docs
```

### CODE REVIEW

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 1]
```diff
@@ -295,24 +295,30 @@ bool TiffDecoder::readHeader()
                 (ncn != 1 && ncn != 3 && ncn != 4)))
                 bpp = 8;
 
+            char depth;
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_tiff.cpp:298]: `int depth;`

It is better to move it below into "case" code blocks:

```
case XXX:
{
    int depth = ...
    .. use depth ...
    break;
}
```

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 2]
```diff
@@ -295,24 +295,30 @@ bool TiffDecoder::readHeader()
...
+            uint16 fmt = SAMPLEFORMAT_UINT;
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_tiff.cpp:299]: fmt -> sample_format

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 3]
```diff
@@ -295,24 +295,30 @@ bool TiffDecoder::readHeader()
...
+            TIFFGetField(tif, TIFFTAG_SAMPLEFORMAT, &fmt);
             int wanted_channels = normalizeChannelsNumber(ncn);
             switch(bpp)
             {
                 case 1:
-                    m_type = CV_MAKETYPE(CV_8U, !isGrayScale ? wanted_channels : 1);
+                    depth = fmt == SAMPLEFORMAT_INT ? CV_8S : CV_8U;
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_tiff.cpp:305]: We need to validate its value first:
```
CV_Check((int)sample_format, sample_format == SAMPLEFORMAT_UINT || sample_format == SAMPLEFORMAT_INT, "");
```

File: modules/imgcodecs/src/grfmt_tiff.cpp [hunk 4]
```diff
@@ -295,24 +295,30 @@ bool TiffDecoder::readHeader()
...
+                        m_type = CV_MAKETYPE(depth, !isGrayScale ? wanted_channels : 1);
                     result = true;
                     break;
                 case 16:
-                    m_type = CV_MAKETYPE(CV_16U, !isGrayScale ? wanted_channels : 1);
+                    depth = fmt == SAMPLEFORMAT_INT ? CV_16S : CV_16U;
+                    m_type = CV_MAKETYPE(depth, !isGrayScale ? wanted_channels : 1);
                     result = true;
                     break;
                 case 32:
```
Comment (alalek) [inline modules/imgcodecs/src/grfmt_tiff.cpp:324]: 32 and 64 cases:
`CV_CheckEQ((int)sample_format, SAMPLEFORMAT_IEEEFP, "");`

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

**Comment by Harvey-Huang** [comment]
thanks for reviewing, I updated the code.

File: modules/imgcodecs/src/grfmt_tiff.cpp (hunk 4 above)
Comment (Harvey-Huang) [reply modules/imgcodecs/src/grfmt_tiff.cpp]: Done, for 32 case, since OpenCV has CV_32S, so 32 case should allow SAMPLEFORMAT_INT?

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:

**Pull request merged**